### PR TITLE
Filter component

### LINKED ISSUE #38: Create radio button & puzzle filter

### PR DESCRIPTION
**Pull request by yuqi88**
- Create radio button component
- Add style to the radio button when toggled
- Create puzzle filter component using the radio button component

Closes #38

### CODE REVIEW

File: src/components/Global/Filter/index.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import React, { useState } from 'react';
+import RadioButton from '../RadioButton';
+import styles from './Filter.module.scss';
+
+const Index = () => {
```
Comment (ketphan02) [inline src/components/Global/Filter/index.tsx:5]: change the component name to Filter

File: src/components/Global/Filter/index.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  const [easyChecked, setEasyChecked] = useState(false);
+  const [mediumChecked, setMediumChecked] = useState(false);
+  const [hardChecked, setHardChecked] = useState(false);
+
```
Comment (ketphan02) [inline src/components/Global/Filter/index.tsx:9]: Just a suggestion
How about having an array of 3 and you will update the element inside that array
```js
const [levelChecker, setLevelChecker] = useState<Array<Boolean>>([false, false, false]);
```
and update will be
```js
const temp = [...levelChecker];
temp[1] = true;
setLevelChecker(temp);
```

Comment (ketphan02) [reply src/components/Global/Filter/index.tsx:9]: not sure if it will run faster but it will have less state to manage

File: src/components/Global/RadioButton/RadioButton.module.scss [hunk 3]
```diff
@@ -0,0 +1,64 @@
+.gone {
+  position: absolute;
+  opacity: 0;
+  cursor: pointer;
+  height: 0;
+  width: 0;
+}
+.normal {
+  display: inline-block;
+  padding: 10px 20px;
```
Comment (ketphan02) [inline src/components/Global/RadioButton/RadioButton.module.scss:10]: change px to rem

File: src/components/Global/RadioButton/RadioButton.module.scss [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  @extend .normal;
+  border-color: #d7222c;
+  color: #d7222c;
+
+  &:hover {
+    text-decoration: underline;
+  }
+}
+//checked style
+//easy:checked ~ label{
```
Comment (ketphan02) [inline src/components/Global/RadioButton/RadioButton.module.scss:46]: you should remove the comments if it's a draft

File: src/components/Global/RadioButton/index.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  name: string /*filter*/;
+  difficulty: string /*Easy>> */;
+  checked: boolean;
+  setChecked: (wasChecked: any) => void;
+};
+const getLabelColor = (difficulty: string, isChecked: boolean) => {
+  // console.log(difficulty + " "+ isChecked);
+  if (difficulty === 'Easy') {
+    if (isChecked) {
+      // console.log(isChecked + " in rdobtn @1@");
```
Comment (ketphan02) [inline src/components/Global/RadioButton/index.tsx:14]: also remove these console.log

File: src/components/Global/RadioButton/index.tsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
+import styles from './RadioButton.module.scss';
+
+type RadioButtonProps = {
+  id: string /*radioEasy*/;
+  name: string /*filter*/;
+  difficulty: string /*Easy>> */;
+  checked: boolean;
+  setChecked: (wasChecked: any) => void;
+};
+const getLabelColor = (difficulty: string, isChecked: boolean) => {
```
Comment (ketphan02) [inline src/components/Global/RadioButton/index.tsx:10]: I think you can improve this function by separating checking "checked" and checking difficulty. You can check the difficulty first then check if checked, then merge them later.

File: src/components/Global/RadioButton/index.tsx [hunk 7]
```diff
@@ -0,0 +1,59 @@
+import styles from './RadioButton.module.scss';
+
+type RadioButtonProps = {
+  id: string /*radioEasy*/;
+  name: string /*filter*/;
+  difficulty: string /*Easy>> */;
```
Comment (ketphan02) [inline src/components/Global/RadioButton/index.tsx:6]: create a type like
```js
type DIFFICULTY = 'hard' | 'medium' | 'easy';
```
and replace string with DIFFICULTY type

File: src/components/Global/RadioButton/index.tsx [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  checked: boolean;
+  setChecked: (wasChecked: any) => void;
```
Comment (ketphan02) [inline src/components/Global/RadioButton/index.tsx:8]: change wasChecked type to boolean

File: styles/login.module.scss [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+}
+.loginPage button{
+  width:  100%;
+  background-color: coral;
+  color: white;
+  border: 0;
+  padding: 8px;
+  border-radius:  8px;
+  cursor: pointer;
+}
```
Comment (ketphan02) [inline styles/login.module.scss:37]: add a new line here

File: src/components/Global/RadioButton/index.tsx (hunk 8 above)
Comment (ketphan02) [reply src/components/Global/RadioButton/index.tsx:8]: also setChecked is a function from useState if you use it like above, it means that the type of this function is not `() => void`, it's `React.Dispatch<React.SetStateAction<boolean>>`

Comment (ketphan02) [reply src/components/Global/RadioButton/index.tsx:8]: This function might have both cases, you can handle this by `React.Dispatch<React.SetStateAction<string>> | (wasChecked: boolean) => void;`

**Review by ketphan02 — CHANGES_REQUESTED** [review]

File: styles/login.module.scss (hunk 9 above)
Comment (yuqi88) [reply styles/login.module.scss:37]: deleted the file in this branch. 
login.tsx is being reviewed in #39

File: src/components/Global/Filter/index.tsx (hunk 2 above)
Comment (yuqi88) [reply src/components/Global/Filter/index.tsx:9]: it's recommended to separate the states, so I'll leave them the way they are.
https://reactjs.org/docs/hooks-faq.html#should-i-use-one-or-many-state-variables

File: src/components/Global/RadioButton/RadioButton.module.scss [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+  height: 0;
+  width: 0;
+}
+.normal {
+  display: inline-block;
+  padding: 0.625rem 1.25rem;
+  margin: 0.625rem 1.25rem;
+  font-family: 'Nunito Sans', serif;
+  //font-size: 1rem;
+  border: 2px solid;
```
Comment (ketphan02) [inline src/components/Global/RadioButton/RadioButton.module.scss:14]: change to rem

File: src/components/Global/RadioButton/RadioButton.module.scss [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+.mediumChecked {
+  @extend .medium;
+  text-decoration: initial;
+  background-color: #3bb2c8;
+  color: white;
+}
+.hardChecked {
+  @extend .hard;
+  text-decoration: initial;
+  background-color: #d7222c;
```
Comment (ketphan02) [inline src/components/Global/RadioButton/RadioButton.module.scss:61]: import the colors from our mixin style

File: src/pages/index.tsx [hunk 12]
```diff
@@ -16,6 +17,7 @@ const Home: React.FC = () => {
     <>
       <h1>Not logged in</h1>
       <button onClick={() => signIn()}>Login</button>
+      <Filter />
```
Comment (ketphan02) [inline src/pages/index.tsx:20]: I think you should remove this

File: src/components/Global/RadioButton/RadioButton.module.scss [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+  text-align: center;
+  cursor: pointer;
+}
+.easy {
+  @extend .normal;
+  border-color: #05c05b;
+  color: #05c05b;
+
+  &:hover {
+    text-decoration: underline;
```
Comment (ketphan02) [inline src/components/Global/RadioButton/RadioButton.module.scss:25]: this is the same for all the button so I think you can move this to the .normal

**Review by ketphan02 — CHANGES_REQUESTED** [review]

File: src/components/Global/Filter/index.tsx [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        />
+        <RadioButton
+          id={'radioHard'}
+          name={'filter'}
+          difficulty={'hard'}
+          checked={hardChecked}
+          setChecked={setHardChecked}
+        />
+      </form>
+      <button
```
Comment (Opeyem1a) [inline src/components/Global/Filter/index.tsx:35]: This button should be a real `<Button>` that says "Search" or something

File: src/components/Global/Filter/index.tsx [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+import styles from './Filter.module.scss';
+
+const Filter = () => {
+  const [easyChecked, setEasyChecked] = useState(false);
+  const [mediumChecked, setMediumChecked] = useState(false);
+  const [hardChecked, setHardChecked] = useState(false);
+
+  return (
+    <div>
+      <form className={styles.filter}>
```
Comment (Opeyem1a) [inline src/components/Global/Filter/index.tsx:12]: The `Filter` component should also contain an `<Input>` and the state of that input method should also be managed here. You'll probably need to merge the updated version (after resolving reviews) of the Input component into this branch again

File: src/components/Global/RadioButton/RadioButton.module.scss [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+  border-color: $secondary-100;
+  color: $secondary-100;
+}
+.hard {
+  @extend .normal;
+  border-color: $red;
+  color: $red;
+}
+.easyChecked {
+  @extend .easy;
```
Comment (Opeyem1a) [inline src/components/Global/RadioButton/RadioButton.module.scss:41]: `@extend` here isn't necessary, you could just add or remove the `easyChecked` class and leave the `easy` class on the component. The current way you have means the `easy` styles are duplicated when an element has both `easy` and `easyChecked`

This applies to `mediumChecked` and `hardChecked`

File: src/components/Global/RadioButton/index.tsx [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+import styles from './RadioButton.module.scss';
+
+type DIFFICULTY = 'easy' | 'medium' | 'hard';
+
+type RadioButtonProps = {
+  id: string /*radioEasy*/;
+  name: string /*filter*/;
+  difficulty: DIFFICULTY;
+  checked: boolean;
+  setChecked: React.Dispatch<React.SetStateAction<boolean>>;
```
Comment (Opeyem1a) [inline src/components/Global/RadioButton/index.tsx:11]: Oh is this the proper way to do this? It looks kinda clunky

File: src/components/Global/RadioButton/index.tsx [hunk 18]
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
+import styles from './RadioButton.module.scss';
+
+type DIFFICULTY = 'easy' | 'medium' | 'hard';
+
+type RadioButtonProps = {
+  id: string /*radioEasy*/;
```
Comment (Opeyem1a) [inline src/components/Global/RadioButton/index.tsx:7]: the comment here and on line 8 should be removed

File: src/components/Global/RadioButton/index.tsx [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+  name: string /*filter*/;
+  difficulty: DIFFICULTY;
+  checked: boolean;
+  setChecked: React.Dispatch<React.SetStateAction<boolean>>;
+};
+const getLabelColor = (difficulty: DIFFICULTY, isChecked: boolean) => {
+  return isChecked ? styles[`${difficulty}Checked`] : styles[`${difficulty}`];
```
Comment (Opeyem1a) [inline src/components/Global/RadioButton/index.tsx:14]: Even though this is a beautiful 1-line code, it creates a dependency between the name of the `DIFFICULTY` types and the names of the classes that I don't think we should have.

So I think the move is:
```ts
if(difficulty === 'easy') return isChecked ? styles.easyChecked : styles.easy;
// repeat for 'medium' and 'hard'
```

File: src/components/Global/RadioButton/index.tsx [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+  return (
+    <div className="radio">
+      <input
+        className={styles.gone}
+        type="radio"
+        id={id}
+        name={name}
+        value={difficulty}
+        onClick={() => {
+          setChecked(!checked);
```
Comment (Opeyem1a) [inline src/components/Global/RadioButton/index.tsx:33]: wait shouldn't it be `setChecked((wasChecked) => !wasChecked);`?

File: src/pages/index.tsx [hunk 21]
```diff
@@ -1,7 +1,7 @@
 import React from 'react';
 import { signIn, signOut, useSession } from 'next-auth/react';
 
-const Home: React.FC = () => {
```
Comment (Opeyem1a) [inline src/pages/index.tsx:4]: Undo this change since it has nothing to do with this branch and will cause conflicts with other branches

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

File: src/components/Global/RadioButton/index.tsx (hunk 17 above)
Comment (yuqi88) [reply src/components/Global/RadioButton/index.tsx:11]: yee. The old one was wrong and more clunky.

File: src/components/Global/RadioButton/index.tsx (hunk 19 above)
Comment (yuqi88) [reply src/components/Global/RadioButton/index.tsx:14]: Is dependency sth we want to reduce? If yes, why is that?
In this case, if we use if-statements, the value of label colour also depends on the value of difficulty, so what's the difference?

File: src/components/Global/Filter/index.tsx [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+        <RadioButton
+          id={'radioEasy'}
+          name={'filter'}
+          difficulty={'easy'}
+          checked={easyChecked}
+          onClick={() => {
+            setEasyChecked(wasChecked => !wasChecked);
+            setFilterFields({
```
Comment (Opeyem1a) [inline src/components/Global/Filter/index.tsx:23]: This is not how we want to do this, ideally we need to set up a `useEffect` hook that listens for changes to `easyChecked`, `mediumChecked`, and `hardChecked` and then uses `setFilterFields()` to match their values. Look into `useEffect()` for implementation details

File: src/components/Global/Filter/index.tsx [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+            setHardChecked(wasChecked => !wasChecked);
+            setFilterFields({
+              easy: easyChecked,
+              medium: mediumChecked,
+              hard: !hardChecked
+            });
+          }}
+        />
+      </form>
+      {/*{console.log("e "+ easyChecked+ " / m"+ mediumChecked+" /h"+hardChecked)}*/}
```
Comment (Opeyem1a) [inline src/components/Global/Filter/index.tsx:59]: Remove unneeded comments

File: src/components/Global/Input/index.tsx [hunk 24]
```diff
@@ -2,20 +2,24 @@ import React, { useState } from 'react';
...
 type InputProps = {
-  type: 'text' | 'password';
+  type: 'text' | 'password' | 'email';
   id: string;
   required: boolean;
   placeholder?: string;
   maxLength?: number;
+  minLength?: number;
   labelText?: string;
+  setInputVal: React.Dispatch<React.SetStateAction<string>>;
```
Comment (Opeyem1a) [inline src/components/Global/Input/index.tsx:12]: Is this type necessary? Does `() => void` not work?

File: src/components/Global/Input/index.tsx [hunk 25]
```diff
@@ -28,8 +32,12 @@ const Input = ({
         required={required}
         placeholder={placeholder}
         maxLength={maxLength}
+        minLength={minLength}
         value={input}
-        onChange={event => setInput(event.target.value)}
+        onChange={event => {
+          setInput(event.target.value);
```
Comment (Opeyem1a) [inline src/components/Global/Input/index.tsx:38]: If you're using the `setInputVal` technique, this component doesn't need to have an internal `input` state variable

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

**Comment by Opeyem1a** [comment]
Sorry I didn't review this earlier @yuqi88, I didn't notice there were more changes - also resolve the merge conflict on this branch so we can hopefully merge it soon

File: src/components/Global/Input/index.tsx (hunk 24 above)
Comment (yuqi88) [reply src/components/Global/Input/index.tsx:12]: - The above way of defining the type is recommended for passing useState function as props though ( ) => void isn't entirely wrong. 
- setInputVal could either take in ( value: string | ( preVal: string) => { } ), and the second type in the return type is a dispatch. 
- Dispatch<SetStateAction<string>> is more generic; thus recommended.

File: src/components/Global/Input/index.tsx (hunk 25 above)
Comment (yuqi88) [reply src/components/Global/Input/index.tsx:38]: The value attribute of the input tag is assigned by the value in the input variable. (line before onChange = ....)
the internal input state is still necessary.

File: src/components/Global/Filter/Filter.module.scss [hunk 26]
```diff
@@ -0,0 +1,5 @@
+.filter {
+  display: flex;
+  justify-content: space-between;
+  width: 19rem;
```
Comment (Opeyem1a) [inline src/components/Global/Filter/Filter.module.scss:4]: why does this have a width at all?

File: src/pages/index.tsx (hunk 12 above)
Comment (Opeyem1a) [reply src/pages/index.tsx:18]: Reminder to do this

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

**Review by Opeyem1a — APPROVED** [review]
